### PR TITLE
Datafile Datatypes

### PR DESCRIPTION
**Pull request by jhdulaney**
## Description
This pull request adds the concept of Datatypes to Open Humans.  It provides a mechanism for site users to add new datatypes to the project.  This mechanism is available to all Open Humans members, though it is only linked to the from the Select Datatypes View, which in turn is only available to Project Coordinators.  In turn, projects immediately after creation are sent to this Select Datatypes View, where they can choose the datatypes they will be using with their project, as well as add new datatypes.

This change also enforces that data_files must have one or more datatypes.

## Related Issue
#981 

## Testing
Added to the automated testing

Manually tested:
Create a project, selected a set of datatypes, added a new datatype
Checked the database directly to ensure datatypes were actually saved
Attempted file upload without specifying a datatype, noted that it failed
Attempted file upload with incorrect datatypes, noted it failed
File upload with full set of project datatypes, succeeded
File upload with subset of project datatypes also succeeded.

### CODE REVIEW

File: private_sharing/api_views.py [hunk 1]
```diff
@@ -301,6 +344,10 @@ class ProjectFileUploadView(ProjectFormBaseView):
     def post(self, request):
         super().post(request)
 
+        # Check datatypes
+        if not self.good_datatypes:
```
Comment (madprime) [inline private_sharing/api_views.py]: This seems like a form validation step -- not something that goes in the view.

File: private_sharing/views.py [hunk 2]
```diff
@@ -676,3 +686,123 @@ def get_object(self, queryset=None):
...
+    def form_valid(self, form):
+        """
+        Form is good, let's save this thing.
+        """
+        ret = super().form_valid(form)
+        if "add-datatype" in form.cleaned_data:
+            # User wants to add a new datatype.  Save the current state of the
+            # selection form.
+            form.cleaned_data.pop("add-datatype")
+            self.request.session["select_category_form"] = form.cleaned_data
```
Comment (madprime) [inline private_sharing/views.py]: it strikes me as very unusual to be using session when filling in forms, I think I'd like to understand why it's being done this way (assuming you're confident this is the correct thing to do)

File: private_sharing/forms.py [hunk 3]
```diff
@@ -393,3 +420,35 @@ class DeleteDataFileForm(forms.Form):
...
+    def clean(self):
+        """
+        Check that something was actually passed, and, if so, populate and return
+        cleaned_data.
+        """
+        super().clean()
+        # Check to see if anything was selected
+        # wants to be on two lines:
+        data = dict(self.data)
+        data.pop("csrfmiddlewaretoken")
```
Comment (madprime) [inline private_sharing/forms.py]: This jumps out to me as something weird/wrong to have to do. I believe normally custom clean functions use `cleaned_data` e.g. for complex validation: https://docs.djangoproject.com/en/2.1/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other

File: private_sharing/views.py (hunk 2 above)
Comment (jhdulaney) [reply private_sharing/views.py]: I'm just using the session to temporarily save form state so that when we return to the form we can reload it.

**Pull request merged**